cmake/EnableCoverage.cmake

```diff
@@ -0,0 +1,118 @@
+		COMMAND ${LCOV} ${LCOV_OPTS} --zerocounters
+		COMMAND ${LCOV} ${LCOV_OPTS} --capture --initial --no-external --output-file ${coverage_dir}/${PROJECT_NAME}_base.info
+		VERBATIM
+		COMMENT "Reseting coverage counters"
+		ALL
+	)
+
+	# Adds the necessary compiler flags to the given target
+	function(enable_coverage_for target)
+		target_compile_options(${target} PRIVATE -fprofile-arcs -ftest-coverage)
+		# apparently the way to set linker flags is through target_link_libraries...
+		target_link_libraries(${target} --coverage)
+		# rebuilding changes the target, thus the coverage will need to be regenerated to be valid
+	endfunction()
+
+	# Sets the given target up for coverage generating, and adds it to the list of targets to call when capturing coverage
+	function(register_coverage_generating_target target)
+		get_target_property(target_type ${target} TYPE)
+		if(target_type STREQUAL EXECUTABLE)
+			enable_coverage_for(${target})
+
+			add_custom_target(coverage_${target} COMMAND ${target} ${ARGN} COMMENT "Running ${target}")
+
+			# this ensures that coverage_reset is executed BEFORE running targets in coverage_capture
+			add_dependencies(coverage_${target} coverage_reset)
+
+			set_property(DIRECTORY ${PROJECT_SOURCE_DIR} APPEND PROPERTY coverage_targets coverage_${target})
+		else()
+			message(AUTHOR_WARNING "register_coverage_generating_target only works with executable targets")
+		endif()
+	endfunction()
+
+	# Creates some common targets required for coverage generation
+	function(finalize_coverage_setup)
+		get_property(coverage_targets DIRECTORY ${PROJECT_SOURCE_DIR} PROPERTY coverage_targets)
+		string(REPLACE ";" " " coverage_targets_joined "${coverage_targets}")
+		message(STATUS "Finalizing coverage setup with these targets: ${coverage_targets_joined}")
+
+		if(use_gcov)
+			set(info_file ${coverage_dir}/${PROJECT_NAME}.info)
+
+			add_custom_target(coverage_capture
+				DEPENDS coverage_reset ${coverage_targets}
+				COMMAND ${LCOV} ${LCOV_OPTS} --capture --no-external --output-file ${info_file}
+				COMMAND ${LCOV} ${LCOV_OPTS} --add-tracefile ${info_file} --add-tracefile ${coverage_dir}/${PROJECT_NAME}_base.info --output-file ${info_file}
+				COMMAND ${LCOV} ${LCOV_OPTS} --remove ${info_file} "moc_*" "${PROJECT_BINARY_DIR}/*" ${ARGN} --output-file ${info_file}
+				BYPRODUCTS ${info_file}
+				VERBATIM
+				COMMENT "Capturing coverage data (this may take some time)"
+			)
+
+			add_custom_target(coverage_html
+				DEPENDS coverage_capture
+				COMMAND ${GENHTML} --demangle-cpp --legend --title ${PROJECT_NAME} --output-directory ${coverage_dir}/html ${info_file}
+				VERBATIM
+				COMMENT "Generating coverage HTML report"
+			)
+
+			set(codecov ${coverage_dir}/codecov.sh)
+			file(DOWNLOAD https://codecov.io/bash ${codecov})
+			add_custom_target(coverage_codecov
+				DEPENDS coverage_capture
+				COMMAND ${codecov} -f ${info_file} -X search -X gcov
+				VERBATIM
+				COMMENT "Uploading coverage data to codecov.io"
+			)
```

Comment:
I dislike the idea of running a unverified executable which is downloaded everytime CMake is run. I would suggest to delegat this download to ExternalProject, with an SHA256 sum for verification.